explorer/interpreter/type_checker.cpp

```diff
@@ -954,199 +982,268 @@ class ConstraintTypeBuilder {
     case Value::Kind::VariableType: {
-      auto it = dict.find(&cast<VariableType>(*type).binding());
-      if (it == dict.end()) {
+      auto it = bindings.args().find(&cast<VariableType>(*type).binding());
+      if (it == bindings.args().end()) {
         return type;
       } else {
         return it->second;
       }
     }
     case Value::Kind::AssociatedConstant: {
       const auto& assoc = cast<AssociatedConstant>(*type);
-      Nonnull<const Value*> base = Substitute(dict, &assoc.base());
-      Nonnull<const Value*> interface = Substitute(dict, &assoc.interface());
-      Nonnull<const Value*> witness = Substitute(dict, &assoc.witness());
+      Nonnull<const Value*> base = Substitute(bindings, &assoc.base());
+      Nonnull<const Value*> interface =
+          Substitute(bindings, &assoc.interface());
+      Nonnull<const Value*> witness = Substitute(bindings, &assoc.witness());
       return arena_->New<AssociatedConstant>(
           base, cast<InterfaceType>(interface), &assoc.constant(),
           cast<Witness>(witness));
     }
     case Value::Kind::TupleValue: {
       std::vector<Nonnull<const Value*>> elts;
       for (const auto& elt : cast<TupleValue>(*type).elements()) {
-        elts.push_back(Substitute(dict, elt));
+        elts.push_back(Substitute(bindings, elt));
       }
       return arena_->New<TupleValue>(elts);
     }
     case Value::Kind::StructType: {
       std::vector<NamedValue> fields;
       for (const auto& [name, value] : cast<StructType>(*type).fields()) {
-        auto new_type = Substitute(dict, value);
+        auto new_type = Substitute(bindings, value);
         fields.push_back({name, new_type});
       }
       return arena_->New<StructType>(std::move(fields));
     }
     case Value::Kind::FunctionType: {
       const auto& fn_type = cast<FunctionType>(*type);
-      std::map<Nonnull<const GenericBinding*>, Nonnull<const Value*>> new_dict(
-          dict);
+      Bindings new_bindings = bindings;
+
+      // Make a new impl binding for a generic binding if needed, and return
+      // its witness.
+      std::vector<Nonnull<const ImplBinding*>> impl_bindings;
+      auto MakeImplBinding = [&](Nonnull<GenericBinding*> new_binding,
+                                 Nonnull<const GenericBinding*> old_binding)
+          -> std::optional<Nonnull<const Witness*>> {
+        if (!old_binding->impl_binding()) {
+          return std::nullopt;
+        }
+        Nonnull<ImplBinding*> impl_binding =
+            arena_->New<ImplBinding>(new_binding->source_loc(), new_binding,
+                                     &new_binding->static_type());
+        impl_binding->set_original(old_binding->impl_binding().value());
+        auto* witness = arena_->New<BindingWitness>(impl_binding);
+        impl_binding->set_symbolic_identity(witness);
+        new_binding->set_impl_binding(impl_binding);
+        impl_bindings.push_back(impl_binding);
+        return witness;
+      };
```

Comment:
I only see this called once, is it benefiting from being a separate lambda?

Note, I view functions and lambdas as different costs -- as a separate function I wouldn't comment, but as a separate lambda I expect it to have a reason for being in this scope. But I also think this code is reaching _serious_ need of refactoring.

Reply:
I think it's worth factoring this out, but maybe it's not worth keeping it as a lambda in this scope. Moved both lambdas to a class that lives outside this scope.